sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.common.policy;
+
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.http.HttpHeaderName;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.BearerTokenAuthenticationPolicy;
+import com.azure.core.util.CoreUtils;
+import reactor.core.publisher.Mono;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * The storage authorization policy which supports challenge.
+ */
+public class StorageBearerTokenChallengeAuthorizationPolicy extends BearerTokenAuthenticationPolicy {
+
+    private static final String DEFAULT_SCOPE = "/.default";
+    private String[] scopes;
+
+    private static final Pattern AUTHENTICATION_CHALLENGE_PATTERN
+        = Pattern.compile(" ((?:\\w+.*?))([a-z]+://[^ ]*)");
+
+    /**
+     * Creates StorageBearerTokenChallengeAuthorizationPolicy.
+     *
+     * @param credential the token credential to authenticate the request
+     * @param scopes the scopes used in credential, using default scopes when empty
+     */
+    public StorageBearerTokenChallengeAuthorizationPolicy(TokenCredential credential, String... scopes) {
+        super(credential, scopes);
+        this.scopes = scopes;
+    }
+
+    @Override
+    public Mono<Void> authorizeRequest(HttpPipelineCallContext context) {
+        return Mono.defer(() -> {
+            String[] scopes = this.scopes;
+            scopes = getScopes(context, scopes);
+            if (scopes == null) {
+                return Mono.empty();
+            } else {
+                return setAuthorizationHeader(context, new TokenRequestContext().addScopes(scopes));
+            }
+        });
+    }
+
+    @Override
+    public void authorizeRequestSync(HttpPipelineCallContext context) {
+        String[] scopes = this.scopes;
+        scopes = getScopes(context, scopes);
+
+        if (scopes != null) {
+            setAuthorizationHeaderSync(context, new TokenRequestContext().addScopes(scopes));
+        }
+    }
+
+    @Override
+    public Mono<Boolean> authorizeRequestOnChallenge(HttpPipelineCallContext context, HttpResponse response) {
+        return Mono.defer(() -> {
+            String authHeader = response.getHeaderValue(HttpHeaderName.WWW_AUTHENTICATE);
+            Map<String, String> challenges = parseChallenges(authHeader);
+
+            String scope = challenges.get("resource_id=");
```

Comment:
Does `authorization_uri` get used for anything?

Reply:
It's not used in this challenge at least - Wasn't sure if it might be necessary in the future so I added the ability to access it just in case.